src/librustc_mir/interpret/validity.rs

```diff
@@ -578,26 +578,36 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValueVisitor<'mir, 'tcx, M>
                     self.path
                 );
             }
-            ty::Array(tys, ..) | ty::Slice(tys)
-                if {
-                    // This optimization applies for types that can hold arbitrary bytes (such as
-                    // integer and floating point types) or for structs or tuples with no fields.
-                    // FIXME(wesleywiser) This logic could be extended further to arbitrary structs
-                    // or tuples made up of integer/floating point types or inhabited ZSTs with no
-                    // padding.
-                    match tys.kind {
-                        ty::Int(..) | ty::Uint(..) | ty::Float(..) => true,
-                        ty::Tuple(tys) if tys.len() == 0 => true,
-                        ty::Adt(adt_def, _)
-                            if adt_def.is_struct() && adt_def.all_fields().next().is_none() =>
-                        {
-                            true
-                        }
-                        _ => false,
+            ty::Array(tys, ..) | ty::Slice(tys) => {
+                // This optimization applies for types that can hold arbitrary bytes (such as
+                // integer and floating point types) or for ZSTS.
+                // Note that we always validate at least one element, even for ZSTs so that no
+                // validation logic is skipped for uninhabited types (which may be zero sized).
+                let elem_layout = self.ecx.layout_of(tys)?;
+                if elem_layout.is_zst() {
+                    // Always validate the first element according to the regular rules.
```

Comment:
I find this comment (and the one above) confusing. What's happening here is:
* We check for ZST.
* All ZST values are the same, so we don't need to check every element of the array/slice, just the first.